### ISSUE TITLE
Protocol does not seem to work if imported

### ISSUE THREAD

**Issue by KelSolaar** [microsoft/pyright#4286]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Hello,

I'm testing `pyright` as replacement of `mypy` and I just stumbled on this problem: It seems like a Protocol class does not work properly if imported.

**To Reproduce**

Given this module: https://github.com/colour-science/colour/blob/develop/colour/algebra/interpolation.py

If at the bottom, I append this code:

```python
from colour.hints import TypeInterpolator

interpolator: TypeInterpolator = LinearInterpolator(
    np.array([-np.inf, np.inf]), np.array([-np.inf, np.inf])
)
```

I get the following error:

```
  /Users/kelsolaar/Documents/Development/colour-science/colour/colour/algebra/interpolation.py:2089:34 - error: Expression of type "LinearInterpolator" cannot be assigned to declared type "TypeInterpolator"
    "LinearInterpolator" is incompatible with "TypeInterpolator" (reportGeneralTypeIssues)
```

However, if I define `TypeInterpolator` within the module, no problem and everything seems fine:

```python
from colour.hints import Protocol

class TypeInterpolator(Protocol):  # noqa: D101
    @property
    def x(self) -> NDArray:
        ...

    @x.setter
    def x(self, value: ArrayLike):
        ...

    @property
    def y(self) -> NDArray:
        ...

    @y.setter
    def y(self, value: ArrayLike):
        ...

    def __init__(self, *args: Any, **kwargs: Any) -> None:  # noqa: D102
        ...  # pragma: no cover

    def __call__(
        self, x: FloatingOrArrayLike
    ) -> FloatingOrNDArray:  # noqa: D102
        ...  # pragma: no cover

interpolator: TypeInterpolator = LinearInterpolator(
    np.array([-np.inf, np.inf]), np.array([-np.inf, np.inf])
)
```

**Expected behavior**

I would expect that the imported Protocol class works like with `Mypy`.

**Screenshots or Code**

See above.

**VS Code extension or command-line**

```bash
pyright --skipunannotated colour/algebra/interpolation.py
```

**Additional context**

Updated `pyproject.toml` file:

```toml
[tool.pyright]
reportMissingImports = false
reportMissingModuleSource = false
reportUnboundVariable = false
reportUnsupportedDunderAll = false
reportUnusedExpression = false
```

**Comment by erictraut** [comment]
The problem is that `hints/__init__.py` is using a dynamic construct to import `Protocol` from either `typing` or `typing_extensions`. 

The `typing_extensions` library provides the abstraction for type features. It implements them if necessary and simply re-exports them from `typing` if not, so there's no need to conditionally import symbols from `typing`. Simply import them from `typing_extensions`.

```python
from typing_extensions import (
    Literal,
    Protocol,
    SupportsIndex,
    TypedDict,
    runtime_checkable,
)
```

If you really want to import symbols dynamically, you can use a `TYPE_CHECKING` conditional:

```python
except ImportError:  # pragma: no cover
    if not TYPE_CHECKING:
        from typing_extensions import (
            Literal,
            Protocol,
            SupportsIndex,
            TypedDict,
            runtime_checkable,
        )
```
